Application connections tech preview (A.K.A. MCP OAuth) (#7104)
elastic/docs-content: deploy-manage/app-connections/oauth-clients.md

type: overview
applies_to:
  serverless: preview
products:
  - id: elasticsearch
  - id: kibana
  - id: cloud-serverless
---

# OAuth for MCP clients [oauth-clients]

The [](/explore-analyze/ai-features/agent-builder/mcp-server.md) supports OAuth 2.1 as a way for MCP clients to authenticate on behalf of a user, alongside [API keys](/explore-analyze/ai-features/agent-builder/mcp-server-api-keys.md).

OAuth is best for interactive, agentic use cases: instead of configuring a static, long-lived API key, a user connects an MCP host such as Claude Desktop and consents in the browser. The MCP client then acts with that user's permissions, using short-lived tokens that the user, a project administrator, or an organization owner can revoke at any time.

:::{note}
<!-- TODO: confirm tool scope for tech preview — PRD said "read-only, ES|QL-based tools only", but QA testing (June 23 2026) shows 22 tools including write operations (delete_stream, create_partition, update_stream, cases, etc.). Remove or correct this sentence once confirmed with Jake Landis / Elena Shostak. -->

> During technical preview, OAuth for the MCP server is available on {{serverless-short}} projects only. Clients are registered within a specific {{serverless-short}} project, so each client is scoped to one project.

Reviewer: > Remove or correct this sentence once 

Permissions follow the user . See https://github.com/elastic/docs-content/pull/7104/changes#r3573009744

> During technical preview, 

Same comment as : https://github.com/elastic/docs-content/pull/7104/changes#r3572909013 ...it is tech preview, but after tech preview the rest of the sentence remains correct.

:::

## OAuth or API keys?

The way that you configure your MCP server depends on your authentication method. Both methods let an MCP client reach the {{agent-builder}} MCP server. To compare them, refer to [MCP server authentication and configuration](/explore-analyze/ai-features/agent-builder/mcp-server.md#mcp-server-authentication).

To learn how to configure your MCP server with API keys, refer to [](/explore-analyze/ai-features/agent-builder/mcp-server-api-keys.md). The rest of this page describes configuring your MCP server with OAuth.

## Key concepts

Understanding these terms makes the setup and management pages easier to follow.

| Term | Definition |
| --- | --- |
| MCP host | The application a user runs that contains MCP clients, such as Claude Desktop or Cursor. Users connect hosts; hosts use clients. |

> | MCP client | The OAuth client registration that holds the credentials (a client ID, and a client secret for confidential clients) your MCP host uses to authenticate to the MCP server. You create one in {{kib}} before connecting a host. |

Reviewer: ```suggestion
| OAuth client | The OAuth client registration that holds the credentials (a client ID, and a client secret for confidential clients) your MCP host uses to authenticate to the MCP server. You create one in {{kib}} before connecting a host. |
```

| MCP server | The interface that exposes {{agent-builder}} tools to MCP hosts. The MCP server is the only resource the OAuth tokens grant access to. This is separate from the [](/explore-analyze/ai-features/agent-builder/tools/mcp-tools.md), which let your agents call external MCP servers. |

> | App connection | The record created when a user consents, linking that user, the MCP client, and the {{serverless-short}} project. A connection is the unit of access and revocation. If two people use the same client ID, each consent creates a separate connection. |

Reviewer: ```suggestion
| App connection | The record created when a user consents, linking that user, the MCP client, and the {{serverless-short}} project. A connection established on top of an OAuth client and is the unit of access and revocation. If two people use the same client ID, each consent creates a separate connection. |
```

## How it works

The OAuth flow follows these steps:

> 1. A user [creates an MCP client](create-oauth-client.md) in {{kib}}, scoped to one {{serverless-short}} project, and receives a client ID and an MCP server URL. These values can be used to [configure MCP hosts](connect-mcp-host.md).

Reviewer: ```suggestion
1. A user [creates an OAuth client](create-oauth-client.md) in {{kib}}, scoped to one {{serverless-short}} project, and receives a client ID and an MCP server URL. These values can be used to [configure MCP hosts](connect-mcp-host.md).
```

> 2. The first time the host needs access, it opens a browser for the user to authenticate and consent. Authentication is always required for consent, even if the user already has an active {{ecloud}} session.

Reviewer: ```suggestion
2. The first time the MCP host (AI agent) needs access, it opens a browser for the user to authenticate and consent. Authentication is always required for consent but may not require explicit login if the user has an active {{ecloud}} session.
```

> 3. On consent, an app connection is created and the client receives tokens. The MCP client presents these to the MCP server, which exchanges them internally to access {{es}} with the user's current permissions.

Reviewer: ```suggestion
3. On consent, an app connection is created and the client receives tokens. The MCP host (AI agent) presents these to the MCP server, which is used to access {{es}} with the approving user's permissions.
```

4. The user, a project administrator, or an organization owner can revoke the connection or the whole client at any time, at the [project](revoke-oauth-client.md) or [organization](manage-app-connections.md) level.

OAuth tokens are accepted only by the {{agent-builder}} MCP server. They don't grant direct access to {{kib}} or {{es}} APIs.

### Sharing a client

> A single client registration can be reused by any number of people. After you create a client, you can share its connection details so that others configure the same client in their own MCP hosts. Each person authenticates and consents separately, which creates a distinct app connection scoped to that person's own permissions. Revoking one person's connection leaves the others active, while revoking the client ends access for everyone.

Reviewer: ```suggestion
A single OAuth client registration can be reused by any number of people. After you create a client, you can share its connection details so that others configure the same client in their own MCP hosts (AI agents). Each person authenticates and consents separately, which creates a distinct app connection scoped to that person's own permissions. Revoking one person's connection leaves the others active, while revoking the client ends access for everyone.
```

### About tokens

Access tokens are short-lived and refreshed automatically in the background, so an active connection keeps working without user action. Refresh is inactivity-based: after 30 days without use, a connection expires and the user must consent again. Because expiry is detected only when a connection is next used, a connection that shows as connected might be idle and not yet revalidated.

### Permissions

> <!-- TODO: confirm tool scope — same as above. If write tools are available in tech preview, remove "read-only, {{esql}}-based" qualifier and update the "can't modify data" statement. -->

Reviewer: > If write tools are available in tech preview

The permissions follow the user. I am not sure if the default tools exposed by agent builder provide write access.  The permission is contrained to the user that user that Authorized the connect and the workflows are contrained by the tools exposed by the Agent builder.

| App connection | The record created when a user consents, linking that user, the MCP client, and the {{serverless-short}} project. A connection is established on top of an OAuth client, and is the unit of access and revocation. If two people use the same client ID, each consent creates a separate connection. |

## How it works

The OAuth flow follows these steps:

1. A user [creates an OAuth client](create-oauth-client.md) in {{kib}}, scoped to one {{serverless-short}} project, and receives a client ID and an MCP server URL. These values can be used to [configure MCP hosts](connect-mcp-host.md).
2. The first time the MCP host (AI agent) needs access, it opens a browser for the user to authenticate and consent. Authentication is always required for consent, but might not require an explicit login if the user already has an active {{ecloud}} session.
3. On consent, an app connection is created and the client receives tokens. The OAuth client presents these to the MCP server, which exchanges them internally to access {{es}} with the user's current permissions.

> 4. The user, a project administrator, or an organization owner can revoke the connection or the whole client at any time, at the [project](revoke-oauth-client.md) or [organization](manage-app-connections.md) level.

Reviewer 2: For organization level, wdyt of linking to the #revoke-connections anchor instead of the whole **Manage app connections** page? It'll be more targeted but the **Revoke connections** section doesn't situate the user and assumes they have read the rest of the page, so might cause more confusion. Just a thought

OAuth tokens are accepted only by the {{agent-builder}} MCP server. They don't grant direct access to {{kib}} or {{es}} APIs.

### Sharing a client

A single OAuth client registration can be reused by any number of people. After you create a client, you can share its connection details so that others configure the same client in their own MCP hosts (AI agents). Each person authenticates and consents separately, which creates a distinct app connection scoped to that person's own permissions. Revoking one person's connection leaves the others active, while revoking the client ends access for everyone.

> ### About tokens

Reviewer 2: ```suggestion
### Access tokens
```
